### ISSUE TITLE
Sympy caching isn't unique

### ISSUE THREAD

**Issue by mloubout** [sympy/sympy#17731]
So we came across this issue as our software (Devito) relies a lot on Sympy symbolic features.

 We rely on Sympy's cache to free our objects that can be very heavy memory-wise (up to O(1e2) Gb per symbol). There are however a few cases where clearing the Sympy cached does not release the memory because subparts of Sympy retain references to these objects in their own "sub-cache" that are to cleared at Sympy's `clear_cache`.

Some of the caches we found are:
```
sympy.polys.rings._ring_cache
sympy.polys.fields._field_cache
sympy.polys.domains.modularinteger._modular_integer_cache
sympy.polys.rootoftools.ComplexRootOf....
```

And there may be other ones as these ones have been found following the `sympy.solve` trace that was/is causing us memory leaks.

**Comment by oscarbenjamin** [comment]
There can be reasons for wanting to have separate caches for some things. Is this issue a request that `clear_cache` should clear all of these caches or that there should be only one cache?

**Comment by mloubout** [comment]
Either of these would be fine, and I don't have a specific preference.

**Comment by oscarbenjamin** [comment]
There's a lot messing around with caches in CRootOf. I dislike the way that CRootOf exposes internal details of caching via the interval attributes. I don't see why it can't just use `cacheit`.

**Comment by asmeurer** [comment]
IIRC some of the polys classes can break if they aren't cached correctly.

**Comment by FabioLuporini** [comment]
>  Is this issue a request that clear_cache should clear all of these caches
I'd say clear_cache should clear all caches

> or that there should be only one cache?
Perhaps just use cacheit, as pretty much everywhere across ``core``?  this would be hitting two birds with one stone -- clear_cache doing its job and a centralised cache

**Comment by oscarbenjamin** [comment]
I don't know if there is a good reason for having separate caches in some of these cases. Otherwise yes, having a single cache would be good.

The main cache is used for many low-level operations and blows through pretty quickly with a size limit of 1000 so it serves a particular function of making things efficient when an operation is repeated in quick succession. My understanding is that it's not so good at e.g. storing an incredibly expensive operation that you definitely don't want to calculate twice even if 1000 other things were calculated in the mean time though.

**Comment by mloubout** [comment]
The cache definitely serves a (good) purpose. The main issue is that even if the user wants to clear the cache, these separate ones such as `poly`  keep references to some of the objects that are asked to be cleared and therefore prevents the `clear-cache()` to do his job.

**Comment by asmeurer** [comment]
Some technical discussion of the polys stuff here https://github.com/sympy/sympy/pull/12088. I would need to look closer at the details again, but I believe it would indeed break things if it were cleared. If a "cache" breaks if it is cleared, it's not really a cache, and should be called something else. 

I tried to refactor these poly classes to not use the singleton pattern in that PR but I was unsuccessful (it was too complicated). If it is resulting in memory leaks it would be worth revisiting it. Can you confirm which of these caches are keeping the references to your objects?

I think even if we can't refactor the polys "cache" thing we can make it a weak dictionary so that they can be garbage collected. The only problems I'm aware of arise from the class needing to be singletonized, so if zero references exist it shouldn't be a problem to delete it and potentially create a new one later. 

For the other caches, I agree they should all use the same global cache. And even if they don't they should always use a LRU cache instance that prevents unbounded memory usage. The fact that the global cache is LRU means that if some code spends a lot of time in interval isolation (for example), the cache entries in the global cache will all stick around in the cache for the duration of the computation, the same as if the interval isolation had its own separate cache. The only downside of a global cache is that things can get crowded and a local computation that makes heavy use of the cache can clear everything else from it. *But*, this is also another way of saying that we require the memory usage to be limited, even if some internal computation occurs that makes heavy use of the cache. The global cache size can also be adjusted pretty easily.

**Comment by asmeurer** [comment]
It looks like some functions do have their own separate cache, for instance, `igcd` (see https://github.com/sympy/sympy/issues/15405 and https://github.com/sympy/sympy/pull/15453). But we should modify clear_cache so that it has a reference to every lru_cache created in the codebase and clears all of them.

**Comment by oscarbenjamin** [comment]
I just read through #12088 and the issues with poly there. To me it looks like another case where attempting to bypass the design constraints of Basic leads us to trouble. I think that any time a Basic subclass defines `__eq__` leads to trouble.

**Comment by asmeurer** [comment]
I don't think the classes are Basic in the polys, but you're still right that getting these things right is difficult, especially if you try to do fancy things.

**Comment by sylee957** [comment]
I think that the immediate solution that the other developers would be expecting is identifying every possible cache usage and providing a single class to clear every caches.

```diff
$ git diff
diff --git a/sympy/core/cache.py b/sympy/core/cache.py
index 809ee639a0..ad33c7f126 100644
--- a/sympy/core/cache.py
+++ b/sympy/core/cache.py
@@ -23,6 +23,32 @@ def print_cache(self):

             print(name, info)

+        # Clear other SymPy cache contents
+        from sympy.combinatorics.free_groups import _free_group_cache
+        print("Free group cache : ", _free_group_cache)
+
+        from sympy.polys.fields import _field_cache
+        print("Field cache : ", _field_cache)
+
+        from sympy.polys.rings import _ring_cache
+        print("Ring cache : ", _ring_cache)
+
+        from sympy.polys.domains.modularinteger import _modular_integer_cache
+        print("Modular integer cache : ", _modular_integer_cache)
+
+        from sympy.integrals.heurisch import _symbols_cache
+        print("Symbols cache : ", _symbols_cache)
+
+        from sympy.integrals.manualintegrate import \
+            _integral_cache, _parts_u_cache
+        print("Integral cache : ", _integral_cache)
+        print("Parts u cache : ", _parts_u_cache)
+
+        from sympy.polys.rootoftools import _reals_cache, _complexes_cache
+        print("ComplexRootsOf Reals cache : ", _reals_cache)
+        print("ComplexRootsOf Complexes cache : ", _complexes_cache)
+
+
     def clear_cache(self):
         """clear cache content"""
         for item in self:
@@ -34,6 +60,31 @@ def clear_cache(self):
                 else:
                     myfunc = myfunc.__wrapped__

+        # Clear other SymPy cache contents
+        from sympy.combinatorics.free_groups import _free_group_cache
+        _free_group_cache.clear()
+
+        from sympy.polys.fields import _field_cache
+        _field_cache.clear()
+
+        from sympy.polys.rings import _ring_cache
+        _ring_cache.clear()
+
+        from sympy.polys.domains.modularinteger import _modular_integer_cache
+        _modular_integer_cache.clear()
+
+        from sympy.integrals.heurisch import _symbols_cache
+        _symbols_cache.clear()
+
+        from sympy.integrals.manualintegrate import \
+            _integral_cache, _parts_u_cache
+        _integral_cache.clear()
+        _parts_u_cache.clear()
+
+        from sympy.polys.rootoftools import ComplexRootOf
+        ComplexRootOf.clear_cache()
+
+
 # global cache registry:
 CACHE = _cache()
 # make clear and print methods available
```

This can be done straightforward like this. Though I think that some more stuff still exists for the `physics.vector`, parsers, sequence, `bernoulli`, or such, but I think that the stuff above are more easier than others because they are using the simple context managers.

Though it may need some tests to see how it works after the cleanups.

**Comment by FabioLuporini** [comment]
@sylee957 yeah, what you wrote is basically the temporary patch I ended up adding to our code, until a fix is available in sympy (we have our own clear_cache that wraps sympy, so it wasn't too difficult)

@asmeurer some questions about what you wrote

> but I believe it would indeed break things if it were cleared.

could you elaborate? I'm now a little worried that if I wipe out those caches from our code, I might have unexpected repercussions, if I get what you wrote right?

>  If a "cache" breaks if it is cleared, it's not really a cache, and should be called something else.

Agreed. It's not a cache, it's a global state, which I bet you don't want to have in sympy . 

>  Can you confirm which of these caches are keeping the references to your objects?

yep. It's the `_ring_cache` and the `_field_cache`. I am not sure about the others -- I only have reproducers for those two

> I don't think the classes are Basic in the polys

I don't remember what exactly, but in my digging yesterday I definitely saw some classes inheriting from Basic

anyway, thanks guys for the quick feedback -- very appreciated !

**Comment by asmeurer** [comment]
@sylee957 I think every LRU cache that is used should go through sympy.core.cache so it can do this automatically.

Also, as I noted, clearing the ring and field caches might break things, so I would be careful with that. 

@FabioLuporini to clarify, from my memory (and I could be wrong about this, so it's better to read through that PR I linked to), the classes assume that only one instance for a given set of arguments ever exists. So if you were to create ring1 (say, `PolyRing([x], QQ)`; I'm not sure if that's the exact syntax), then the next time you call `PolyRing([x], QQ)` it will return the same object from the "cache" (I put cache in quotes because it's not really a cache. A real cache shouldn't break things if you clear it). If you were to clear the cache and create that same object, any preexisting `PolyRing([x], QQ)` would not compare equal to it. The reason is a bit technical (it's because one of the attributes checked in `__eq__` is a dynamically created class, whose metaclass doesn't implement the proper `__eq__`). 

It's unclear exactly how this could break things. Having duplicate objects floating around that are supposed to be equal but don't actually compare equal can lead to all kinds of weird bugs (just look up the sorts of weird things that can happen when you use `reload()`, for example). 

My suggested fix is to use a weak dictionary. As a workaround for your code you should be able to go through the dict and clear any element that has a `sys.getrefcount` of 2 (one for the cache entry and one for `sys.getrefcount` itself). A more correct fix would be to refactor the polys classes to not use crazy metaclass stuff, but that is a lot more complicated obviously.
